src/Suite.ts

```diff
@@ -1,52 +1,106 @@
+  storyStart: (storyName: string, componentName: string) => void;
+  storyPass: (storyName: string, componentName: string, result: Result.Result, elapsedTime: number) => void;
+  storyFail: (storyName: string, componentName: string, result: Result.Result, elapsedTime: number) => void;
+  storySkip: (storyName: string, componentName: string) => void;
+  storyError: (storyName: string, componentName: string, error: Error) => void;
+  suiteFinish: (browser: string, numPass: number, numFail: number, numSkip: number, elapsedTime: number) => void;
+}
+
+/**
+ * Event emitter representing a run of the test suite.
+ */
+export type SuiteEmitter = Emitter<SuiteEvents>;
+
+/**
+ * Find Storybook stories and run Axe on each one. Returns an event emitter that emits events when
+ * components and stories are processed.
+ */
+export function run(options: Options): SuiteEmitter {
+  const emitter = createEmitter<SuiteEvents>();
+
+  defer(async () => {
+    const suiteStartTime = Date.now();
+    let numPass = 0;
+    let numFail = 0;
+    let numSkip = 0;
 
-    each(storiesByComponent, (stories, componentTitle) => {
-      const nameMatches = options.pattern.test(componentTitle);
-      const describeFn = nameMatches ? describe : describe.skip;
-      const describeName = nameMatches ? componentTitle : `[skipped] ${componentTitle}`;
+    emitter.emit('suiteStart', options.browser);
 
-      describeFn(describeName, () => {
-        stories.forEach((story) => {
-          const testFn = ProcessedStory.isEnabled(story) ? it : it.skip;
+    // Get the Storybook stories.
+    const testBrowser = await TestBrowser.create(options);
+    const page = await TestBrowser.createPage(testBrowser, options);
+    const stories = await TestBrowser.getStories(page);
+    const storiesByComponent = groupBy(stories, 'componentName');
+    const storiesAndComponents = Object.entries(storiesByComponent);
 
-          testFn(story.name, async function () {
-            const result = await Result.fromPage(page, story);
+    try {
+      // Iterate each component.
+      for (const [componentName, stories] of storiesAndComponents) {
+        const shouldComponentRun = options.pattern.test(componentName);
+
+        if (shouldComponentRun) {
+          emitter.emit('componentStart', componentName);
+        } else {
+          emitter.emit('componentSkip', componentName);
+        }
+
+        // Iterate each story in this component.
+        for (const story of stories) {
+          const storyStartTime = Date.now();
+
+          if (!shouldComponentRun || !ProcessedStory.isEnabled(story)) {
+            numSkip += 1;
+            emitter.emit('storySkip', story.name, componentName);
+            continue;
+          }
+
+          emitter.emit('storyStart', story.name, componentName);
+
+          try {
+            // Detect any Axe violations for this story.
+            const result = await pTimeout(Result.fromPage(page, story), options.timeout);
```

Comment:
@anniehu4 , I've added a timeout here. This will throw an error that gets caught by the `catch` below.